### PR TITLE
SOLR-16878: Add JUL Log4J manager for lucene logs

### PR DESCRIPTION
**Pull request by HoustonPutman**
https://issues.apache.org/jira/browse/SOLR-16878

This can be tested relatively easily, **using Java 20**

```
$ ./gradlew dev
$ cd solr/packaging/build/dev
$ ./bin/solr start -e films
```

In the solr logs (not the console logs), you will see:

```
INFO  (coreLoadExecutor-10-thread-1) [ x:films] o.a.l.s.MemorySegmentIndexInputProvider Using MemorySegmentIndexInput with Java 20; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
```

For the new Vector logs, after introducing the option in Solr in #1758, no logging is done by default. But we can use the C1 compiler, which makes Lucene log that it can't use the Panama Vector API.

```
$ ./gradlew dev
$ cd solr/packaging/build/dev
$ ./bin/solr start -e films -a "-XX:TieredStopAtLevel=1"
```
In the solr logs (again, not the console logs), you will see:

```
WARN  (qtp1027495011-41) [ x:films] o.a.l.u.VectorUtilProvider C2 compiler is disabled; Java vector incubator API can't be enabled
```

### CODE REVIEW

**Comment by uschindler** [comment]
It should also log if it is enabled: https://github.com/apache/lucene/blob/d03c8f16d9d4f87ce24d603e4a9b4a7806f9b533/lucene/core/src/java20/org/apache/lucene/internal/vectorization/PanamaVectorizationProvider.java#L74

Maybe the issue is that VectorUtil is not used on IndexReader or IndexWriter init.

Tip: Elasticsearch triggers class initialization using `MethodHandles.lookup().ensureClassInitialited(VectorUtil.class)`. Same for MMapDirectory. This enables warnings and similar directly on startup.

I am in cinema, so will later look into it.

**Comment by HoustonPutman** [comment]
Unfortunately `MethodHandles.lookup().ensureInitialited(VectorUtil.class)` is only allowed in Java 15+, and Solr is on Java 11.

We could always just do a quick:

```java
VectorUtil.dotProduct(new byte[0], new byte[0]);
```

This works, so will probably commit separately. Do we need to be worried about the security manager stuff, or will the lucene hack work around it for us?

**Comment by uschindler** [comment]
The alternative is: `Class.forName(VectorUtil.class.getName())`

Security Manager is fine for us as Lucene wraps it. Solr has no components that run with zero privileges like Elasticsearch script engine.

File: solr/bin/solr [hunk 1]
```diff
@@ -1581,7 +1581,8 @@ if [ "${SOLR_HOME:0:${#EXAMPLE_DIR}}" = "$EXAMPLE_DIR" ]; then
   SOLR_LOGS_DIR="$SOLR_HOME/../logs"
 fi
 
-LOG4J_CONFIG=()
+# Set the logging manager by default, so that Lucene JUL logs are included with Solr logs.
+LOG4J_CONFIG=("-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager")
```
Comment (uschindler) [inline solr/bin/solr:1585]: It looks like this is misisng for Windows `solr.cmd`.

Small question: I hope this is an internal variable that is not overwritten by user code in `solr.in.sh`?

**Review by uschindler — CHANGES_REQUESTED** [review]

**Comment by uschindler** [comment]
To make sure that all logs are written on early startup, we should add this code to the early startup phase after logging is fully configured:

```java
Stream.of(MMapDirectory.class, VectorUtil.class).forEach(cls -> {
 try {
  Class.forName(cls.getName());
 } catch (ReflectiveOperationException re) {
  throw new SolrException(...)
 }
});
```

Maybe add a TODO to use `MethodHandles.lookup().ensureClassInitialized` in Solr 10 w/ Java 17:

```java
final var lookup = MethodHandles.lookup();
Stream.of(MMapDirectory.class, VectorUtil.class).forEach(cls -> {
 try {
  lookup.ensureInitialized(cls);
 } catch (ReflectiveOperationException re) {
  throw new SolrException(...)
 }
});
```

File: solr/bin/solr (hunk 1 above)
Comment (HoustonPutman) [reply solr/bin/solr:1585]: No, `solr.in.sh` cannot override this. But good call on windows, will add that in.

Comment (HoustonPutman) [reply solr/bin/solr:1585]: Ok it should work for windows too, but it does it in a slightly different way. Shouldn't be a big deal.

**Review by risdenk — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java [hunk 2]
```diff
@@ -226,6 +229,21 @@ public void init(ServletContext servletContext) {
...
+          .forEach(
+              cls -> {
+                try {
+                  Class.forName(cls.getName());
+                } catch (ReflectiveOperationException re) {
+                  throw new SolrException(
+                      ErrorCode.SERVER_ERROR, "Could not load Lucene class: " + cls.getName());
+                }
+              });
+      VectorUtil.dotProduct(new byte[0], new byte[0]);
```
Comment (uschindler) [inline solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java:245]: If the class is initialized (what the previous code does), this call is obsolete.

**Review by uschindler — COMMENTED** [review]
Works for me with Windows:

Java 20:
```
2023-07-10 18:30:15.272 INFO  (main) [] o.e.j.s.Server jetty-10.0.15; built: 2023-04-11T17:25:14.480Z; git: 68017dbd00236bb7e187330d7585a059610f661d; jvm 20+36-2344
2023-07-10 18:30:16.726 WARN  (main) [] o.e.j.u.DeprecationWarning Using @Deprecated Class org.eclipse.jetty.servlet.listener.ELContextCleaner
2023-07-10 18:30:16.781 INFO  (main) [] o.a.s.s.CoreContainerProvider Using logger factory org.apache.logging.slf4j.Log4jLoggerFactory
2023-07-10 18:30:16.798 INFO  (main) [] o.a.s.s.CoreContainerProvider  ___      _       Welcome to Apache Solr™ version 10.0.0-SNAPSHOT 8c3b093601c8a42a7d8170a054d34e2dabe09219 [snapshot build, details omitted]
2023-07-10 18:30:16.798 INFO  (main) [] o.a.s.s.CoreContainerProvider / __| ___| |_ _   Starting in standalone mode on port 8983
2023-07-10 18:30:16.798 INFO  (main) [] o.a.s.s.CoreContainerProvider \__ \/ _ \ | '_|  Install dir: C:\Users\Uwe Schindler\Projects\lucene\solr\solr\packaging\build\solr-10.0.0-SNAPSHOT-slim
2023-07-10 18:30:16.799 INFO  (main) [] o.a.s.s.CoreContainerProvider |___/\___/_|_|    Start time: 2023-07-10T18:30:16.799817100Z
2023-07-10 18:30:16.805 INFO  (main) [] o.a.s.s.CoreContainerProvider Solr started with "-XX:+CrashOnOutOfMemoryError" that will crash on any OutOfMemoryError exception. The cause of the OOME will be logged in the crash file at the following path: C:\Users\Uwe Schindler\Projects\lucene\solr\solr\packaging\build\solr-10.0.0-SNAPSHOT-slim\example\techproducts\solr\..\logs\jvm_crash_35668.log
2023-07-10 18:30:16.842 INFO  (main) [] o.a.l.s.MemorySegmentIndexInputProvider Using MemorySegmentIndexInput with Java 20; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
2023-07-10 18:30:16.960 INFO  (main) [] o.a.l.u.VectorUtilPanamaProvider Java vector incubator API enabled; uses preferredBitSize=256
```

Java 21:

```
2023-07-10 18:33:09.880 INFO  (main) [] o.e.j.s.Server jetty-10.0.15; built: 2023-04-11T17:25:14.480Z; git: 68017dbd00236bb7e187330d7585a059610f661d; jvm 21-ea+28-2377
2023-07-10 18:33:11.551 WARN  (main) [] o.e.j.u.DeprecationWarning Using @Deprecated Class org.eclipse.jetty.servlet.listener.ELContextCleaner
2023-07-10 18:33:11.614 INFO  (main) [] o.a.s.s.CoreContainerProvider Using logger factory org.apache.logging.slf4j.Log4jLoggerFactory
2023-07-10 18:33:11.631 INFO  (main) [] o.a.s.s.CoreContainerProvider  ___      _       Welcome to Apache Solr™ version 10.0.0-SNAPSHOT 8c3b093601c8a42a7d8170a054d34e2dabe09219 [snapshot build, details omitted]
2023-07-10 18:33:11.631 INFO  (main) [] o.a.s.s.CoreContainerProvider / __| ___| |_ _   Starting in standalone mode on port 8983
2023-07-10 18:33:11.631 INFO  (main) [] o.a.s.s.CoreContainerProvider \__ \/ _ \ | '_|  Install dir: C:\Users\Uwe Schindler\Projects\lucene\solr\solr\packaging\build\solr-10.0.0-SNAPSHOT-slim
2023-07-10 18:33:11.632 INFO  (main) [] o.a.s.s.CoreContainerProvider |___/\___/_|_|    Start time: 2023-07-10T18:33:11.632405500Z
2023-07-10 18:33:11.637 INFO  (main) [] o.a.s.s.CoreContainerProvider Solr started with "-XX:+CrashOnOutOfMemoryError" that will crash on any OutOfMemoryError exception. The cause of the OOME will be logged in the crash file at the following path: C:\Users\Uwe Schindler\Projects\lucene\solr\solr\packaging\build\solr-10.0.0-SNAPSHOT-slim\example\techproducts\solr\..\logs\jvm_crash_10432.log
2023-07-10 18:33:11.672 INFO  (main) [] o.a.l.s.MemorySegmentIndexInputProvider Using MemorySegmentIndexInput with Java 21; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
2023-07-10 18:33:11.845 INFO  (main) [] o.a.l.u.VectorUtilPanamaProvider Java vector incubator API enabled; uses preferredBitSize=256
```

My question and request to change: Why do you have the obsolete Vector dotProduct. `Class#forName` by default initializes the class (you see this also with MMaüDirectory, the message comes directly before that one).

So this is obsolete and should be removed: `VectorUtil.dotProduct(new byte[0], new byte[0]);`
(I am also not sure if it is useful to calculate the dot prod of empty vectors because it does not test anything.

**Comment by uschindler** [comment]
It also shows the log levels here: 
image

**Comment by uschindler** [comment]
Oh it is merged already! Mhm, maybe remove the obsolete line, otherwise all is fine!

File: solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java (hunk 2 above)
Comment (HoustonPutman) [reply solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java:245]: My bad, I thought I had deleted that. Will remove.

**Comment by uschindler** [comment]
Looks fine on main branch!

I just noticed: we still have "-XX:CompileCommand=exclude,com.github.benmanes.caffeine.cache.BoundedLocalCache::put" in the command line options. Wasn't this bug fixed? Before opening another issue I ask here....

**Comment by uschindler** [comment]
Ignore this, I asked question on #1090.

**Pull request merged**